### PR TITLE
[MRG+2] FIX: always decode bytes as utf-8 in AFM headers

### LINKED ISSUE #3517: Issue with non-ascii paths in font look up
I posted in #3487 and following the suggestion of @tacaswell , I create a separate issue, for easier bug tracking:

Since upgrading to 1.4.0, I get the same exception.
I'm on Mac, Mavericks, latest version. anaconda bundle.

I'm not even sure it's about matplotlib, but it crashes when it gets to font_manager.

This is the relevant output for the exception:

``` python
---> 18 import matplotlib.pyplot as plt
     19
     20 from copy import deepcopy

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/pyplot.py in <module>()
     25
     26 import matplotlib
---> 27 import matplotlib.colorbar
     28 from matplotlib import style
     29 from matplotlib import _pylab_helpers, interactive

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/colorbar.py in <module>()
     32 import matplotlib.artist as martist
     33 import matplotlib.cbook as cbook
---> 34 import matplotlib.collections as collections
     35 import matplotlib.colors as colors
     36 import matplotlib.contour as contour

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/collections.py in <module>()
     25 import matplotlib.artist as artist
     26 from matplotlib.artist import allow_rasterization
---> 27 import matplotlib.backend_bases as backend_bases
     28 import matplotlib.path as mpath
     29 from matplotlib import _path

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/backend_bases.py in <module>()
     54
     55 import matplotlib.tight_bbox as tight_bbox
---> 56 import matplotlib.textpath as textpath
     57 from matplotlib.path import Path
     58 from matplotlib.cbook import mplDeprecation

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/textpath.py in <module>()
     17 from matplotlib.path import Path
     18 from matplotlib import rcParams
---> 19 import matplotlib.font_manager as font_manager
     20 from matplotlib.ft2font import FT2Font, KERNING_DEFAULT, LOAD_NO_HINTING
     21 from matplotlib.ft2font import LOAD_TARGET_LIGHT

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/font_manager.py in <module>()
   1410                 verbose.report("Using fontManager instance from %s" % _fmcache)
   1411         except:
-> 1412             _rebuild()
   1413     else:
   1414         _rebuild()

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/font_manager.py in _rebuild()
   1395     def _rebuild():
   1396         global fontManager
-> 1397         fontManager = FontManager()
   1398         if _fmcache:
   1399             pickle_dump(fontManager, _fmcache)

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/font_manager.py in __init__(self, size, weight)
   1035         #  Load TrueType fonts and create font dictionary.
   1036
-> 1037         self.ttffiles = findSystemFonts(paths) + findSystemFonts()
   1038         self.defaultFamily = {
   1039             'ttf': 'Bitstream Vera Sans',

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/font_manager.py in findSystemFonts(fontpaths, fontext)
    320                     fontfiles[f] = 1
    321
--> 322             for f in get_fontconfig_fonts(fontext):
    323                 fontfiles[f] = 1
    324

/Users/ifmichael/anaconda/lib/python2.7/site-packages/matplotlib/font_manager.py in get_fontconfig_fonts(fontext)
    272         pipe = subprocess.Popen(['fc-list', '--format=%{file}\\n'],
    273                                 stdout=subprocess.PIPE,
--> 274                                 stderr=subprocess.PIPE)
    275         output = pipe.communicate()[0]
    276     except (OSError, IOError):

/Users/ifmichael/anaconda/lib/python2.7/subprocess.pyc in __init__(self, args, bufsize, executable, stdin, stdout, stderr, preexec_fn, close_fds, shell, cwd, env, universal_newlines, startupinfo, creationflags)
    708                                 p2cread, p2cwrite,
    709                                 c2pread, c2pwrite,
--> 710                                 errread, errwrite)
    711         except Exception:
    712             # Preserve original exception in case os.close raises.

/Users/ifmichael/anaconda/lib/python2.7/subprocess.pyc in _execute_child(self, args, executable, preexec_fn, close_fds, cwd, env, universal_newlines, startupinfo, creationflags, shell, to_close, p2cread, p2cwrite, c2pread, c2pwrite, errread, errwrite)
   1325                         raise
   1326                 child_exception = pickle.loads(data)
-> 1327                 raise child_exception
   1328
   1329

UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 0: ordinal not in range(128)
```

Right now I downgraded to 1.3.1, and everything got back to normal.

### PR DESCRIPTION
**Pull request by tacaswell**
closes #3517

### CODE REVIEW

**Comment by tacaswell** [comment]
This is a 'can-not-import-bug', but has been here a long time, the critical flag is to make sure we talk about it.

To test this I think we need to find a debian font which has a non-ascii font family name.

**Comment by NelleV** [comment]
I think just checking that afm._to_str works on non ascii path is enough.

**Comment by NelleV** [comment]
Can I still review this PR? (cause it looks very fine to me :+1: for merging!)

File: lib/matplotlib/tests/test_afm.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import matplotlib.afm as afm
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_afm.py]: This file is missing our standard `__future__` imports.

Comment (NelleV) [inline lib/matplotlib/tests/test_afm.py]: But the future import is useless for this file.

Comment (tacaswell) [inline lib/matplotlib/tests/test_afm.py]: actually, it might be harmful due to the 'unicode_literals'...

Comment (QuLogic) [inline lib/matplotlib/tests/test_afm.py]: In the future, if anyone adds new tests here, then they won't have to worry about adding in these imports if necessary. Easy consistency is important for tests.

Comment (NelleV) [inline lib/matplotlib/tests/test_afm.py]: This is the kind of practices that break static checkers such as pyflakes with very little benefit considering a forgotten import should break the test.

Comment (QuLogic) [inline lib/matplotlib/tests/test_afm.py]: I've never seen a static checker complain about `__future__` imports; how would they even know if you used, e.g., `division` with types that were integers?

This is about consistent Python 2/3 behaviour, not some random import.

File: lib/matplotlib/tests/test_afm.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+
+import matplotlib.afm as afm
+import six
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_afm.py]: It appears that other changes mean this can be removed?

File: lib/matplotlib/tests/test_afm.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+
+def test_nonascii_str():
+    # This tests that we also decode bytes as utf-8 properly.
+    # Else, font files with non ascii caracters fail to load.
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_afm.py]: caracters -> characters

File: lib/matplotlib/tests/test_afm.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+from __future__ import (absolute_import, division, print_function,
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_afm.py]: Needs an encoding comment; this fails on Python 2.

Comment (NelleV) [inline lib/matplotlib/tests/test_afm.py]: Fixed!

**Pull request merged**